nav2_mission_executor/src/mission_executor.cpp

```diff
@@ -12,107 +12,75 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include <string>
-#include <memory>
-#include <exception>
 #include "nav2_mission_executor/mission_executor.hpp"
+#include "nav2_mission_executor/execute_mission_behavior_tree.hpp"
 
-using namespace std::chrono_literals;
 using nav2_tasks::TaskStatus;
 
-namespace nav2_mission_execution
+namespace nav2_mission_executor
 {
 
 MissionExecutor::MissionExecutor()
 : nav2_tasks::ExecuteMissionTaskServer("ExecuteMissionNode")
 {
-  RCLCPP_INFO(get_logger(), "Initializing MissionExecutor.");
-  navTaskClient_ = std::make_unique<nav2_tasks::NavigateToPoseTaskClient>(this);
-
-  if (!navTaskClient_->waitForServer(nav2_tasks::defaultServerTimeout)) {
-    RCLCPP_ERROR(get_logger(), "MissionExecutor: NavigateToPoseTaskServer not running!");
-    throw std::runtime_error("MissionExecutor: NavigateToPoseTaskServer not running");
-  }
-
   goal_sub_ = create_subscription<geometry_msgs::msg::PoseStamped>("move_base_simple/goal",
       std::bind(&MissionExecutor::onGoalPoseReceived, this, std::placeholders::_1));
 
   plan_pub_ = create_publisher<nav2_msgs::msg::MissionPlan>(
     "ExecuteMissionTask_command");
 }
 
-MissionExecutor::~MissionExecutor()
+TaskStatus MissionExecutor::execute(const nav2_tasks::ExecuteMissionCommand::SharedPtr command)
 {
-  RCLCPP_INFO(get_logger(), "Shutting down MissionExecutor");
+  RCLCPP_INFO(get_logger(), "Starting mission execution");
+
+  // Create and run the behavior tree for this mission
+  ExecuteMissionBehaviorTree bt(shared_from_this());
+  TaskStatus result =
+    bt.run(command->mission_plan, std::bind(&MissionExecutor::cancelRequested, this));
+
+  RCLCPP_INFO(get_logger(), "Completed mission execution: %d", result);
+  return result;
 }
 
-void
-MissionExecutor::onGoalPoseReceived(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
+// Besides receving a mission plan on the command topic, another way to initiate the mission
+// is to respond to the goal message sent from rviz. In this case, we'll receive the incoming
+// message, dynamically create a mission plan with a single NavigateToPose, then publish the plan
+void MissionExecutor::onGoalPoseReceived(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
 {
-  goal_pose_ = msg;
-
+  RCLCPP_DEBUG(get_logger(), "Received goal pose message");
+
+  // Get a reference for convenience
+  geometry_msgs::msg::Pose & p = msg->pose;
+
+  // Compose the args for the NavigateToPose action
+  std::stringstream args;
+  args << "position=\"" <<
+    p.position.x << ";" << p.position.y << ";" << p.position.z << "\"" <<
+    " orientation=\"" <<
+    p.orientation.x << ";" << p.orientation.y << ";" << p.orientation.z << ";" <<
+    p.orientation.w << "\"";
+
+  // Put it all together, trying to make the XML somewhat readable here
+  std::stringstream ss;

```

Comment:
Can you rename 'ss' to something descriptive of the data like navigate_mission?

Reply:
Yes, fixed.